File: SHJ3/apps/ai/src/shj3_ai/observability/tracing.py
"""Process-wide OpenTelemetry SDK bootstrap and inbound instrumentation.

Why this exists
----------------
deployment.md §13.1: OpenTelemetry in both runtimes, W3C ``traceparent``
propagated end to end. On this side that means initialising a
``TracerProvider`` (with an OTLP exporter when one is configured) and
instrumenting the FastAPI app with ``opentelemetry-instrumentation-fastapi``,
which extracts an inbound ``traceparent`` automatically through
OpenTelemetry's own context propagation — no hand-rolled header parsing,
matching the web tier's use of ``@opentelemetry/api``'s propagation API
(``apps/web/src/modules/platform/observability/trace-context.ts``).

Why failures here are swallowed, not raised
--------------------------------------------
``SHJ3_OTEL_EXPORTER_OTLP_ENDPOINT`` is empty in ``.env.example`` — local
development runs with no collector — and even in a deployed environment an
unreachable collector must not become a citizen-facing incident. Tracing is
observability, never a request-path dependency: ``SHJ3_OTEL_EXPORTER_OTLP_
ENDPOINT`` is not among the variables deployment.md §4.1 requires the process
to refuse to boot over, so ``configure_tracing`` never raises. A
configuration problem is logged once and the process keeps serving traffic.
"""

from __future__ import annotations

import logging
import os
from typing import TYPE_CHECKING

from opentelemetry import trace
from opentelemetry.exporter.otlp.proto.http.trace_exporter import OTLPSpanExporter
from opentelemetry.instrumentation.fastapi import FastAPIInstrumentor
from opentelemetry.sdk.resources import Resource
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor
from opentelemetry.sdk.trace.id_generator import RandomIdGenerator
from opentelemetry.semconv.resource import ResourceAttributes

if TYPE_CHECKING:
    from fastapi import FastAPI

logger = logging.getLogger(__name__)

_SERVICE_NAME = "shj3-ai"

# See the equivalent comment in the web tier's tracing.ts: deployment.md documents the
# unprefixed OTEL_EXPORTER_OTLP_ENDPOINT, but every process-configuration variable this
# codebase actually reads carries the SHJ3_ prefix (.env.example), which is the source of
# truth a developer's machine boots from.
_OTLP_ENDPOINT_VAR = "SHJ3_OTEL_EXPORTER_OTLP_ENDPOINT"

_configured = False


def configure_tracing() -> None:
    """Initialise the process-wide ``TracerProvider``.

    Idempotent and never raises — see the module docstring. Safe to call more
    than once (test setup, a future ``shj3_ai.entrypoint`` calling it
    alongside :func:`instrument_fastapi_app`): a second call is a no-op.
    """
    global _configured
    if _configured:
        return

    endpoint = os.environ.get(_OTLP_ENDPOINT_VAR, "").strip()

    # Everything, including exporter construction — not just registration — is inside
    # the try block. A malformed endpoint string rejected by the exporter's own
    # constructor is exactly the kind of configuration problem this function exists to
    # survive.
    try:
        resource = Resource.create(
            {
                ResourceAttributes.SERVICE_NAME: _SERVICE_NAME,
                ResourceAttributes.SERVICE_VERSION: os.environ.get("SHJ3_RELEASE_VERSION", "dev"),
                ResourceAttributes.DEPLOYMENT_ENVIRONMENT: os.environ.get(
                    "SHJ3_ENVIRONMENT", "development"
                ),
            }
        )
        provider = TracerProvider(resource=resource)

        if endpoint:
            # OTLPSpanExporter's `endpoint` is the whole request URL, not just the
            # collector host, so the signal-specific suffix is added here — the same
            # choice the web side makes in tracing.ts, for the same reason: deployment.md
            # documents the bare collector address.
            exporter = OTLPSpanExporter(endpoint=f"{endpoint.rstrip('/')}/v1/traces")
            provider.add_span_processor(BatchSpanProcessor(exporter))
        else:
            logger.info(
                "%s is not set; spans are recorded but not exported. This is the expected "
                "development default (.env.example).",
                _OTLP_ENDPOINT_VAR,
            )

        trace.set_tracer_provider(provider)
    except Exception:
        # Deliberately broad — see the module docstring. A tracing-setup failure of any
        # kind must degrade to "no export", never take the process down with it.
        logger.exception(
            "shj3-ai: failed to start the OpenTelemetry SDK; continuing without tracing."
        )
        return

    _configured = True


def reset_tracing_for_testing() -> None:
    """Drop the "already configured" flag so a test can call ``configure_tracing`` again.

    Mirrors ``platform/config.ts``'s ``resetConfigForTesting`` on the web side. Production
    tracing is configure-once-per-process; this exists purely so different tests can
    exercise :func:`configure_tracing`'s full body (an unset endpoint, an unreachable one,
    …) instead of every test after the first hitting the idempotency guard. It does not
    reset OpenTelemetry's own global tracer-provider registration — a second
    ``set_tracer_provider`` call against an already-configured global is itself a safe,
    logged no-op, which is what makes that safe to leave alone.
    """
    global _configured
    _configured = False


def instrument_fastapi_app(app: FastAPI) -> None:
    """Instrument ``app`` so inbound ``traceparent`` headers are extracted automatically.

    ``FastAPIInstrumentor`` reads the ASGI scope's headers through
    ``opentelemetry.propagate`` — the globally registered propagator, W3C
    tracecontext by default — and starts a server span whose trace id is the
    propagated one when a valid ``traceparent`` is present, or a freshly
    generated root trace id otherwise. That extract-or-generate behaviour is
    deployment.md §13.1 rule 1, achieved with library code rather than a
    hand-rolled header parse.
    """
    FastAPIInstrumentor.instrument_app(app)


def current_trace_id() -> str:
    """The active span's trace id, as 32 lowercase hex characters.

    Falls back to a freshly generated, valid W3C trace id when there is no
    active span — called before :func:`instrument_fastapi_app` has run, or
    outside a request entirely (a worker job establishing its own trace, per
    deployment.md §13.1 rule 5). Never raises: this backs a FastAPI
    dependency on the request path (``adapters/inbound/trace_context.py``),
    and a tracing bug must not become a 500.
    """
    span_context = trace.get_current_span().get_span_context()
    if span_context.is_valid:
        return trace.format_trace_id(span_context.trace_id)
    return trace.format_trace_id(RandomIdGenerator().generate_trace_id())

File: SHJ3/apps/web/src/modules/platform/observability/tracing.ts
/**
 * OpenTelemetry SDK bootstrap for `shj3-web`.
 *
 * architecture.md §10: "One trace id spans web → ai → tool call, which is
 * what makes B14 tab 3 real rather than decorative." deployment.md §13.1
 * makes the mechanism a contract rather than a hope: OpenTelemetry in both
 * runtimes, W3C `traceparent` propagated end to end.
 *
 * This module owns every vendor import the tracing story needs on this side
 * — `@opentelemetry/sdk-node`, its OTLP exporter, and the resource helpers.
 * Everything downstream (`trace-context.ts`'s inbound extraction and
 * outbound injection, in turn used by the auth middleware and the AI client)
 * talks to `@opentelemetry/api` only — the same thin surface both runtimes'
 * SDKs speak — so neither hand-parses the other's header (deployment.md
 * §13.1 rule 1: propagation lives in one place, not per call site).
 *
 * ## Why this can never be a request-path dependency
 *
 * `SHJ3_OTEL_EXPORTER_OTLP_ENDPOINT` is deliberately empty in `.env.example`
 * — local development has no collector running — and even in a deployed
 * environment the collector can be unreachable without that being a citizen
 * -facing incident. `startTracing()` therefore never throws: a configuration
 * or export problem is logged once and the process keeps serving traffic
 * without spans leaving the building, rather than refusing to boot the way
 * `platform/config.ts` refuses to boot over a missing *required* secret.
 * Observability is a nice-to-have on the request path; it is not on the list
 * of things deployment.md §4.1 requires the process to die for.
 */

import { NodeSDK } from "@opentelemetry/sdk-node";
import type { NodeSDKConfiguration } from "@opentelemetry/sdk-node";
import { OTLPTraceExporter } from "@opentelemetry/exporter-trace-otlp-http";
import { resourceFromAttributes } from "@opentelemetry/resources";
import { ATTR_SERVICE_NAME, ATTR_SERVICE_VERSION } from "@opentelemetry/semantic-conventions";

/** Matches `OTEL_RESOURCE_ATTRIBUTES`'s `service.name` in deployment.md §4.2 #68. */
const SERVICE_NAME = "shj3-web";

/**
 * deployment.md documents the OTLP endpoint as the unprefixed
 * `OTEL_EXPORTER_OTLP_ENDPOINT` (#67), but `.env.example` and every other
 * process-configuration variable in this codebase carries the `SHJ3_` prefix
 * (`SHJ3_SQL_URL`, `SHJ3_AI_BASE_URL`, …) rather than relying on a vendor
 * SDK's own auto-detected names. This module follows the codebase's actual
 * convention — `.env.example` is the source of truth for what a developer's
 * machine reads — over the doc's naming, which is the one place the two
 * disagree.
 */
const OTLP_ENDPOINT_VAR = "SHJ3_OTEL_EXPORTER_OTLP_ENDPOINT";

/**
 * `deployment.environment` (deployment.md §4.2 #68's `OTEL_RESOURCE_ATTRIBUTES`
 * example). This is an OpenTelemetry *experimental* resource semantic
 * convention, so it is not re-exported from `@opentelemetry/semantic-
 * conventions`'s stable entry point the way `ATTR_SERVICE_NAME` is — hence
 * the literal here rather than a second, incubating import.
 */
const ATTR_DEPLOYMENT_ENVIRONMENT = "deployment.environment";

let sdk: NodeSDK | null = null;

/**
 * Start the process-wide `TracerProvider` and register the W3C propagator.
 *
 * Idempotent and safe to call from more than one place: Next.js can load
 * this module several times per process (route handlers, middleware, the
 * `instrumentation.ts` boot hook, and `trace-context.ts`'s own lazy
 * self-heal on first use), and a second `NodeSDK.start()` would otherwise
 * try to re-register a global tracer provider and propagator, which the API
 * tolerates but logs a warning for on every call.
 */
export function startTracing(): void {
  if (sdk !== null) return;

  const endpoint = process.env[OTLP_ENDPOINT_VAR]?.trim();

  // Everything, including construction — not just `start()` — is inside the try block.
  // An exporter or SDK constructor rejecting a malformed endpoint string is exactly the
  // kind of configuration problem this function exists to survive (module docstring).
  try {
    const resource = resourceFromAttributes({
      [ATTR_SERVICE_NAME]: SERVICE_NAME,
      [ATTR_SERVICE_VERSION]: process.env.SHJ3_RELEASE_VERSION ?? "dev",
      [ATTR_DEPLOYMENT_ENVIRONMENT]: process.env.SHJ3_ENVIRONMENT ?? "development",
    });

    const configuration: Partial<NodeSDKConfiguration> = { resource };
    if (endpoint) {
      // The exporter's `url` option is the *whole* request URL, not just the collector
      // host — unlike the auto env-detected path, which appends the signal-specific
      // suffix itself. deployment.md documents the bare collector address
      // (`http://otel-collector…:4317`), so that suffix is added here to keep the
      // documented variable meaning "collector address" rather than "full traces
      // endpoint".
      configuration.traceExporter = new OTLPTraceExporter({
        url: `${endpoint.replace(/\/+$/, "")}/v1/traces`,
      });
    }

    const candidate = new NodeSDK(configuration);
    candidate.start();
    sdk = candidate;
  } catch (error) {
    // Never throw out of here — see the module docstring. A failed start leaves no
    // propagator registered, which degrades `trace-context.ts` to "every request starts
    // a fresh trace" rather than crashing anything.
    console.error(
      `[observability] failed to start the OpenTelemetry SDK for ${SERVICE_NAME}; ` +
        "continuing without tracing.",
      error,
    );
    return;
  }

  if (!endpoint) {
    console.info(
      `[observability] ${OTLP_ENDPOINT_VAR} is not set; spans are recorded but not exported. ` +
        "This is the expected development default (.env.example).",
    );
  }
}

/**
 * Flush buffered spans and stop the SDK.
 *
 * Best-effort: a failed flush on shutdown must not turn a clean process exit
 * into a crashed one, so the error is logged and swallowed rather than
 * propagated.
 */
export async function shutdownTracing(): Promise<void> {
  const active = sdk;
  sdk = null;
  if (!active) return;

  try {
    await active.shutdown();
  } catch (error) {
    console.error("[observability] failed to flush spans on shutdown", error);
  }
}

/**
 * Drop the cached SDK handle so a test can call `startTracing()` again.
 *
 * Mirrors `platform/config.ts`'s `resetConfigForTesting` — production tracing is
 * start-once-per-process, and this exists purely so different tests can exercise
 * `startTracing()`'s full body (an unset endpoint, an unreachable one, …) instead of every
 * test after the first hitting the idempotency guard. It does not unregister the global
 * OpenTelemetry API state, only this module's own "already started" flag; re-calling
 * `NodeSDK.start()` against an already-registered global provider is itself a safe no-op
 * (a swallowed diagnostic, not a throw), which is what makes that safe to leave alone.
 */
export function resetTracingForTesting(): void {
  sdk = null;
}
